squashfs-mount.c

```diff
@@ -128,19 +118,148 @@ int main(int argc, char **argv) {
+  return strcmp(((const mount_entry_t *)p1)->mountpoint,
+                ((const mount_entry_t *)p2)->mountpoint);
+}
+
+/// split by `:` and convert to abspath, sort by mountpoint
+static mount_entry_t *parse_mount_entries(char **argv, int argc) {
+  // TODO `:` in argv get overwritten by `\0` in this function, is this OK?
+  mount_entry_t *mount_entries = malloc(sizeof(mount_entry_t) * argc);
+
+  for (int i = 0; i < argc; ++i) {
+    char *mnt, *file;
+    if (!(file = strtok(argv[i], ":")) || !(mnt = strtok(NULL, ":"))) {
+      errx(EXIT_FAILURE, "invalid format %s", argv[i]);
+    } else {
+      if (strtok(NULL, ":")) {
+        // expect file:mountpoint, strtok must return NULL when called once more
+        errx(EXIT_FAILURE, "invalid format %s", argv[i]);
+      }
+    }
+
+    strcpy(mount_entries[i].squashfs_file, file);
+    strcpy(mount_entries[i].mountpoint, mnt);
+
+    // convert to absolute paths (if needed)
+    // absolute paths are skipped, since we allow to do nested mounts (for given
+    // absolute), they won't be resolvable via realpath, since in general
+    // non-existent outside the image
+    if ((file[0] != '/') &&
+        realpath(file, mount_entries[i].squashfs_file) == NULL) {
+      errx(EXIT_FAILURE, "Failed to obtain realpath of %s, error: %s", file,
+           strerror(errno));
+    }
+    if ((mnt[0] != '/') && realpath(mnt, mount_entries[i].mountpoint) == NULL) {
+      errx(EXIT_FAILURE, "Failed to obtain realpath of %s, error: %s", mnt,
+           strerror(errno));
+    }
+  }
+
+  // sort by mountpoint
+  qsort(mount_entries, argc, sizeof(mount_entry_t), compare_mountpoint);
+
+  return mount_entries;
+}
+
+int main(int argc, char **argv) {
+  char **fwd_argv;
+  mount_entry_t *mount_entries;
+  uid_t uid = getuid();
 
-  // set env variables allowing to detect the spank plugin if
-  // squashfs file has been mounted before calling srun,etc.
-  if (setenv(ENV_MOUNT_FILE, squashfs_file, 1 /* overwite if exists */) ||
-      setenv(ENV_MOUNT_POINT, mountpoint, 1 /* overwrite if exists */)) {
+  char *program = argv[0];
+
+  argv++;
+  argc--;
+
+  int positional_args = 0;
+  // Early exit for -h, --help, -v, --version.
+  for (int i = 0; i < argc; ++i, positional_args++) {
+    char const *arg = argv[i];
+    // Skip positional args.
+    if (arg[0] != '-' || arg[1] == '\0') {
+      continue;
+    }
+
+    // Early exit on -h, --help, -v, --version
+    if (strcmp(arg, "--") == 0) {
+      break;
+    }
+    if (strcmp(arg, "-h") == 0 || strcmp(arg, "--help") == 0)
+      help(program);
+    if (strcmp(arg, "-v") == 0 || strcmp(arg, "--version") == 0) {
+      puts(VERSION);
+      exit(EXIT_SUCCESS);
+    }
+    // Error on unrecognized flags.
+    errx(EXIT_FAILURE, "Unknown flag %s", argv[i]);
+  }
+
+  if (argc <= positional_args + 1) {
+    exit_with_error("no command given");
+  }
+
+  fwd_argv = argv + (positional_args + 1);
+  // if no mountpoints given, run command directly
```

Comment:
Should this be an error, or at least a warning message?

It doesn't make much sense to call squashfs-mount without any mount points/images

Reply:
The idea was that when `squashfs-mount` is called in a parametrized script, it will still work if no image is given. I've added a warning message.